### ISSUE TITLE
apiserver errors should include reason

### ISSUE THREAD

**Issue by discordianfish** [kubernetes/kubernetes#50622]
/kind feature

**What happened**:
Spend too much time figuring out that my apiserver didn't start because the etcd DNS name couldn't be resolved. I got the following error:
```
Error: error waiting for etcd connection: timed out waiting for the condition
```
And even `--v=10` didn't provide any more context. This was very misleading and made me assume the issue is on application level.

**What you expected to happen**:
I expected the apiserver error to tell me that DNS resolution wasn't working.

**How to reproduce it (as minimally and precisely as possible)**:
I used the bootkube manifest but just starting the apiserver with `--etcd-servers` pointing to a nonexisting name.

**Comment by discordianfish** [comment]
@kubernetes/sig-cluster-ops

**Comment by xiangpengzhao** [comment]
maybe 
/sig api-machinery
also.

**Comment by Cryptophobia** [comment]
@discordianfish : How did you ultimately fix this issue? We are getting the same message when deploying with kops and we think its a DNS issue but we keep getting this error in the kube-apiserver.log file.

**Comment by discordianfish** [comment]
@Cryptophobia It's been a "special" setup, but in general you can verify whether it's a DNS issue by using nslookup or dig.

**Comment by Cryptophobia** [comment]
@discordianfish : It turns out instanceGroups are very important for configuring DNS Route53 and etcd during cluster configuration with kops. If those instanceGroups are not defined correctly **(particularly when there are even number of master or multiple groups of masters into one availability zone in a region)**, etcd servers will not start and master nodes will not check in. This is definitely something that is not very well documented in the kops documentation.

**Comment by discordianfish** [comment]
@Cryptophobia Ah, you should fill an issue with kops then. This issue here is mostly about saving time by pointing you into the right direction.

**Comment by Cryptophobia** [comment]
Okay, I'll add another issue to the 784 issues already open. :laughing:  :+1:

**Comment by ntfrnzn** [comment]
I'm commenting here only to give my +1 to the underlying issue, that the error message emitted by the apiserver is not sufficient to begin debugging related problems.

I'm seeing the same error message, my experimental configuration is quite different, but it would be more pleasing if the apiserver told me clearly _why_ it cannot talk to an etcd cluster, _where_ it thinks that etcd cluster is, and so on.

**Comment by Cryptophobia** [comment]
Agreed. Better error messages and more error details going up to the api layer errors so we know what kubernetes apiserver is actually trying to.

**Comment by Cryptophobia** [comment]
This would be a really nice feature. I wish I knew more Go so I could help out.

**Comment by holygits** [comment]
I've opened a PR for this. Could use a review

**Comment by nikhita** [comment]
> I've opened a PR for this. 

@holygits thanks!! 

I'll remove the help-wanted label on this. :)

/remove-help

**Comment by aaronchall** [comment]
Will this issue ever be addressed?
